widget/entry-with-completion.go

```diff
@@ -0,0 +1,140 @@
+	widget.Entry
+	popupMenu     *widget.PopUp
+	navigableList *navigableList
+	items         []string
+	pause         bool
+}
+
+// NewCompletionEntry creates a new CompletionEntry which creates a popup menu that responds to keystrokes to navigate through the items without losing the editing ability of the text input.
+func NewCompletionEntry(options []string) *CompletionEntry {
+	c := &CompletionEntry{items: options}
+	c.ExtendBaseWidget(c)
+	return c
+}
+
+// ShowCompletion displays the completion menu
+func (c *CompletionEntry) ShowCompletion() {
+	if c.pause {
+		return
+	}
+
+	if c.navigableList == nil {
+		c.navigableList = newNavigableList(c.items, &c.Entry, c.setTextFromMenu, c.HideCompletion)
+	}
+	holder := fyne.CurrentApp().Driver().CanvasForObject(c)
+	pos := fyne.CurrentApp().Driver().AbsolutePositionForObject(c)
+
+	if c.popupMenu == nil {
+		c.popupMenu = widget.NewPopUp(c.navigableList, holder)
+	}
+	max := fyne.Min(10, float32(len(c.items)))
+	c.popupMenu.Resize(fyne.NewSize(c.Entry.Size().Width, c.popupMenu.MinSize().Height*max-3*theme.Padding()))
+	c.popupMenu.ShowAtPosition(fyne.Position{X: pos.X, Y: pos.Y + c.Size().Height})
+	holder.Focus(c.navigableList)
+}
+
+// HideCompletion hides the completion menu.
+func (c *CompletionEntry) HideCompletion() {
+	if c.popupMenu != nil {
+		c.popupMenu.Hide()
+	}
+}
+
+// SetOptions set the completion list with itemList.
+func (c *CompletionEntry) SetOptions(itemList []string) {
+	c.items = itemList
+	if c.navigableList != nil {
+		c.navigableList.SetOptions(itemList)
+	}
+}
+
+// Prevent the menu to open when the user validate value from the menu.
+func (c *CompletionEntry) setTextFromMenu(s string) {
+	c.pause = true
+	c.Entry.SetText(s)
+	c.Entry.Refresh()
+	c.pause = false
+	c.popupMenu.Hide()
+}
+
+type navigableList struct {
+	*widget.List
+	data            binding.StringList
+	entry           *widget.Entry
+	selected        int
+	setTextFromMenu func(string)
+	hide            func()
+	navigating      bool
+}
+
+func newNavigableList(items []string, entry *widget.Entry, setTextFromMenu func(string), hide func()) *navigableList {
+	n := &navigableList{entry: entry, selected: -1, setTextFromMenu: setTextFromMenu, hide: hide}
+	n.data = binding.BindStringList(&items)
+
+	n.List = widget.NewListWithData(n.data, func() fyne.CanvasObject {
+		return widget.NewLabel("")
+	}, func(i binding.DataItem, o fyne.CanvasObject) {
+		o.(*widget.Label).Bind(i.(binding.String))
+	})
+
+	n.OnSelected = func(id int) {
+		if !n.navigating {
+			item, _ := n.data.GetItem(id)
+			val, _ := item.(binding.String).Get()
+			setTextFromMenu(val)
+		}
+		n.navigating = false
+	}
+
+	return n
+}
+
+func (n *navigableList) SetOptions(items []string) {
+	n.Unselect(n.selected)
+	n.data.Set(items)
+	n.selected = -1
+}
+
+func (n *navigableList) TypedKey(event *fyne.KeyEvent) {
+	n.entry.TypedKey(event)
+	switch event.Name {
+	case fyne.KeyDown:
+		if n.selected < n.data.Length()-1 {
+			n.selected++
```

Comment:
Maybe should be better, if the n.selected resets to 0, when it goes until the end:
```go
if n.selected < n.data.Length()-1 {
	n.selected++
} else {
	n.selected = 0
}
```

Reply:
I'm not sure. If the user scroll down and press Down, while the List won't autoscroll, it will lose the selection cursor. I would change this when the List from Fyne does the autoscroll, isn't it ?

Reply:
Actually... You're right. The scroll problem would be resolved later but the common completion list behavior is to "rotate" in the list.

Reply:
That's right. However I wonder if the completion list really need the scroll feature. For example, in chrome the input autocompletion does not scroll it always show a fixed length list of options. Also, scrolling a large list of options isn't a good user experience, is it? 🤔